### PR TITLE
AArch64 Fix getFrameRegs assertion during indirect fixup

### PR DESCRIPTION
**Pull request by jim-saxman**
On AA4ch64 plaforms when calling getFrameRegs() and an indirect
fixup is found, the RIP needs to be present at the correct
offset. It was not, because the destructor was compiled with a
tail-call. It needs to have a standard stack frame.

By disabling the tail-call via the compiler flag on all destructors
in g_destructors[], we can ensure that all destructors for AArch64
are walked correctly.

This bug occurred sometimes in OSS Mediawiki after a few minutes.
We tried to create a simplified unit test but were unsuccessful.

This patch introduces no new unit test failures for either DebugOpt
or Release build types on AArch64 platforms and doesn't affect x64
or PPC64 destructors.

### CODE REVIEW

**Comment by jim-saxman** [comment]
@mxw Hi Max, can you lease review this PR.
@swalk-cavium Are you still seeing the errors in test/quick/*dtor.php unti tests, and does this patch help?

**Comment by apinski-cavium** [comment]
You just use __aarch64__ instead __AARCH64EL__ (not that matters but it is just a style thing).
Also it seems better to do something like:

inline void ArrayData::release() noexcept {
  assert(hasExactlyOneRef());
  g_array_funcs.release[kind()](this);
  AARCH64_WALKABLE_FRAME ();
 }

And have AARCH64_WALKABLE_FRAME (); expand to something like:
__asm__("":::"memory");

I am not a fan of the optimize attribute in this case :).

**Comment by mxw** [comment]
I think `__aarch64__` is more consistent with what we're using elsewhere.

I chatted with @dave-estes offline who seems to like the memory barrier approach.  You can continue ifdeffing that macro to do nothing on x86-64.

File: hphp/runtime/base/array-data-defs.h [hunk 1]
```diff
@@ -214,7 +214,9 @@ inline Variant ArrayData::getKey(ssize_t pos) const {
 
 inline void ArrayData::release() noexcept {
   assert(hasExactlyOneRef());
-  return g_array_funcs.release[kind()](this);
+  g_array_funcs.release[kind()](this);
+  AARCH64_WALKABLE_FRAME();
+  return;
```
Comment (mxw) [inline hphp/runtime/base/array-data-defs.h]: Just delete this return; this function returns `void`.

File: hphp/runtime/base/mixed-array.cpp [hunk 2]
```diff
@@ -468,7 +468,6 @@ ArrayData* MixedArray::MakeDictFromAPC(const APCArray* apc) {
 
 //=============================================================================
 // Destruction
-
```
Comment (mxw) [inline hphp/runtime/base/mixed-array.cpp]: Put this back?

**Comment by swalk-cavium** [comment]
@jim-saxman - I ran a full MOP on this change.  

These two tests now pass with these option sets
hphp/test/run -v -m jit -a '-vEval.JitPGO=1'
hphp/test/run -v -m jit -r -a '-vEval.JitPGO=1'
    hphp/test/quick/setg-dtor.php
    hphp/test/quick/sets-dtor.php

hphp/test/slow/async/simple_meth.php passes all option sets.

A bunch of failures have reappeared with this option set
    hphp/test/run -v -m jit -r -a '-vEval.JitPGO=1'
    hphp/test/zend/good/ext/spl/tests/iterator_002.php
    hphp/test/zend/good/ext/spl/tests/iterator_033.php
    hphp/test/zend/good/ext/spl/tests/iterator_040.php
    hphp/test/zend/good/ext/spl/tests/iterator_048.php
    hphp/test/zend/good/ext/spl/tests/observer_001.php
    hphp/test/zend/good/ext/spl/tests/observer_002.php
    hphp/test/zend/good/ext/spl/tests/RecursiveCallbackFilterIteratorTest.php

I think I saw them last between fixes for relocation.

**Comment by jim-saxman** [comment]
Hi @swalk-cavium, Thanks for running tests on this patch. I re-ran all unit tests in all configurations and observed the following:

For PGO with and without repoAuth:
hphp/test/quick/setg-dtor.php
hphp/test/quick/sets-dtor.php
pass both before and after this patch.

hphp/test/slow/async/simple_meth.php
still fails with -r -m pgo

php/test/zend/good/ext/spl/tests/iterator_002.php
hphp/test/zend/good/ext/spl/tests/iterator_033.php
hphp/test/zend/good/ext/spl/tests/iterator_040.php
hphp/test/zend/good/ext/spl/tests/iterator_048.php
hphp/test/zend/good/ext/spl/tests/observer_001.php
hphp/test/zend/good/ext/spl/tests/observer_002.php
hphp/test/zend/good/ext/spl/tests/RecursiveCallbackFilterIteratorTest.php
never failed both before and after this patch, for -r -m pgo.

I performed these tests on both CentOS with GCC 6.2.1 and Ubuntu with GCC 5.4.

**Comment by swalk-cavium** [comment]
Hi @jim-saxman, I'm wondering if one of my machines needs a firmware upgrade.

**Comment by vielmetti** [comment]
Hi @swalk-cavium @jim-saxman - just did a build and test cycle and am seeing these errors:

HipHop VM 3.20.0-dev (rel)
Compiler: heads/master-0-g26a2f8c2795726e89dd636c16b2e1e6316dee7df
Repo schema: c9ccdb2b68ae4d61f3c1c0232b4b2426ff39c26e
gcc version 5.4.0 20160609 (Ubuntu/Linaro 5.4.0-6ubuntu1~16.04.4) 

```
test/slow/ext_icu/uspoof.php
test/slow/ext_image/1794.php
test/slow/ext_posix/ext_posix.php
test/slow/ext_spl_file/1795.php
test/slow/ir_inlining/1.php
test/quick/setg-dtor.php
test/quick/sets-dtor.php
```
System under test is a 96-core Packet Type 2A server, with Cavium ThunderX, using Ubuntu 16.04. Rather than noting a new issue, since the set*-dtor failures seem to be related, I'll just report here.

**Comment by swalk-cavium** [comment]
Hi Ed,


We have yet to have an all green run on the regression tests.

Our Jenkins runs 6 different option sets.


Which options did you use below?


The uspoof ones depend on which version of the internationalization

library is used.  This one always fails our CI testing.  It has something

to do with libicui18n.so


The ext_posix.php test assumes that there is a 'wheel'  group.  Try

sudo groupadd wheel


I was not able to get 1794.php or 1795.php to fail.


I think these are related to the PR in the subject.

ir_inlining/1.php, setg-dtor.php and sets-dtor.php


Theobroma Systems makes their CI available here.  One option

set, multiple OSs.

http://hhvm-arm64.theobroma-systems.com/


Steve Walk




________________________________

**Comment by vielmetti** [comment]
After applying this patch all of test/quick passes ("SHIP IT").

All of test/slow passes except test/slow/ext_icu/uspoof.php which is described by #3925.

Looking good!

**Pull request closed without merging**